### DISCUSSION TITLE
Splink 4.0.0: How to debug value error in training.py?

### DISCUSSION THREAD

**Discussion by aalexandersson** [moj-analytical-services/splink#2316] [Q&A] [answered]
On my test data, this code worked in Splink 3.9.15:
```
deterministic_rules = [
    "l.ssn = r.ssn"
]
linker.estimate_probability_two_random_records_match(deterministic_rules, recall=0.85)
```
The output was:
```
Probability two random records match is estimated to be 1.42e-06.
This means that amongst all possible pairwise record comparisons, one in 705,054.21 are
expected.
```

But this equivalent "training" code fails in Splink 4.0.0:
```
deterministic_rules = [
     block_on("ssn"),
]
linker.training.estimate_probability_two_random_records_match(deterministic_rules, recall=0.85)
```

The error message is:
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "V:\Testing\Splink_2025\.venv\Lib\site-packages\splink\internals\linker_components\training.py", line 120, in estimate_probability_two_random_records_match
    raise ValueError(
ValueError: Deterministic matching rules led to more observed matches than is consistent with supplied recall. With these rules, recall must be at least -0.00.
```

This is the affected code in **training.py**:
```
        if num_observed_matches > num_total_comparisons * recall:
            raise ValueError(
                f"Deterministic matching rules led to more "
                f"observed matches than is consistent with supplied recall. "
                f"With these rules, recall must be at least "
                f"{num_observed_matches/num_total_comparisons:,.2f}."
            )
```

I arbitrarily tried several other "deterministic_rules", but the error message is the same in Splink 4.0.0. How can I easily calculate `num_observed_matches` and `num_total_comparisons` to be able to debug the error? I tried to re-use code from lines 103-119 but failed.

**Comment by RobinL** [comment]
Thanks for the report.

I'm not sure.  The error message suggests there are no observed matches  since that appears to be the only way you can get a value of `-0.0` from `f"{num_observed_matches/num_total_comparisons:,.2f}."` but it's a bit unclear to me why the 
`if num_observed_matches > num_total_comparisons * recall:` if statement has been triggered.

I would start by adding these lines 

```
        print(f"num_observed_matches: {num_observed_matches}")
        print(f"num_total_comparisons: {num_total_comparisons}")
        print(f"recall: {recall}")
        print(f"estimated matches: {num_total_comparisons * recall}")
```

to here:

https://github.com/moj-analytical-services/splink/blob/89eae40e9355d9067975ff84615baaf18a6558c2/splink/internals/linker_components/training.py#L117



Here's a testing script I wrote to try and produce a reprex (minimal reproducible example), but I can't reproduce the error and it seems to give the right answer:

<details>
<summary>
testing script
</summary>

```python
import pandas as pd

import splink.comparison_library as cl
from splink import DuckDBAPI, Linker, SettingsCreator, block_on, splink_datasets

db_api = DuckDBAPI()

# fmt: off
data_1 = [
    {"unique_id":1, "first_name": "Alice", "surname": "Smith", "dob": "1970-01-01", "city": "London"},
    {"unique_id":2, "first_name": "Bob", "surname": "Jones", "dob": "1984-02-02", "city": "London"},
    {"unique_id":3, "first_name": "Dave", "surname": "Smith", "dob": "1970-01-02", "city": "London"},
]

data_2 = [
    {"unique_id":4, "first_name": "Alice", "surname": "Smith", "dob": "1971-05-05", "city": "London"},
    {"unique_id":5, "first_name": "Bob", "surname": "Jones", "dob": "1990-10-10", "city": "London"},
]
# fmt: on

df_1 = pd.DataFrame(data_1)
df_2 = pd.DataFrame(data_2)
df_all = pd.concat([df_1, df_2])

settings = SettingsCreator(
    link_type="dedupe_only",
    comparisons=[
        cl.ExactMatch("first_name"),
        cl.ExactMatch("surname"),
        cl.ExactMatch("dob"),
        cl.ExactMatch("city"),
    ],
)

linker = Linker(df_all, settings, db_api)

deterministic_rules = [
    block_on("first_name"),
]
linker.training.estimate_probability_two_random_records_match(
    deterministic_rules,
    recall=1.0,
)


settings = SettingsCreator(
    link_type="link_only",
    comparisons=[
        cl.ExactMatch("first_name"),
        cl.ExactMatch("surname"),
        cl.ExactMatch("dob"),
        cl.ExactMatch("city"),
    ],
)

linker = Linker([df_1, df_2], settings, db_api)

linker.training.estimate_probability_two_random_records_match(
    deterministic_rules,
    recall=1.0,
)

settings = SettingsCreator(
    link_type="link_and_dedupe",
    comparisons=[
        cl.ExactMatch("first_name"),
        cl.ExactMatch("surname"),
        cl.ExactMatch("dob"),
        cl.ExactMatch("city"),
    ],
)

linker = Linker([df_1, df_2], settings, db_api)

linker.training.estimate_probability_two_random_records_match(
    deterministic_rules,
    recall=1.0,
)



settings = SettingsCreator(
    link_type="link_and_dedupe",
    comparisons=[
        cl.ExactMatch("first_name"),
        cl.ExactMatch("surname"),
        cl.ExactMatch("dob"),
        cl.ExactMatch("city"),
    ],
)

linker = Linker([df_1, df_2], settings, db_api)

deterministic_rules = [
    block_on("dob"),
]
linker.training.estimate_probability_two_random_records_match(
    deterministic_rules,
    recall=0.85,
)


settings = SettingsCreator(
    link_type="link_and_dedupe",
    comparisons=[
        cl.ExactMatch("first_name"),
        cl.ExactMatch("surname"),
        cl.ExactMatch("dob"),
        cl.ExactMatch("city"),
    ],
)

linker = Linker([df_1, df_2], settings, db_api)

deterministic_rules = [
    block_on("city"),
]
linker.training.estimate_probability_two_random_records_match(
    deterministic_rules,
    recall=0.85,
)

```

</details>

**Reply by aalexandersson** [reply]
Thanks for the help. Your test script ran as I expected; the last section gave an error but with a value of `1.00` instead of `-0.0`.

After adding the suggested lines, here is the output for my test data:
```
num_observed_matches: 541561
num_total_comparisons: -2147483648
recall: 0.85
estimated matches: -1825361100.8
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "V:\Testing\Splink_2025\.venv\Lib\site-packages\splink\internals\linker_components\training.py", line 124, in estimate_probability_two_random_records_match
    raise ValueError(
ValueError: Deterministic matching rules led to more observed matches than is consistent with supplied recall. With these rules, recall must be at least -0.00.
```

How can `num_total_comparisons` be negative?

**Reply by RobinL** [reply]
Aha - `-2147483648` is the minimum value for an unsigned integer, so that is a bit of a tip off as to the source of the problem.  I'll dig a bit deeper to see what's going on, but it looks likely to be a bug to me

**Reply by aalexandersson** [reply]
Thanks for the help. I agree it looks likely to be a bug in Splink 4. I use Windows 10 with the default DuckDB backend. Here is some representative artificial data from one of the two datasets. I can provide more sample data privately if you would like. 
```
df2.head().ssn
```
```
0    449632668
1    349518166
2          NaN
3    553530413
4    718020331
Name: ssn, dtype: object
```

**Reply by RobinL** [reply]
I'm just having a look and I found this PR that makes relevant code changes, specifically the use of bigint:

https://github.com/moj-analytical-services/splink/pull/2263

Could I double check you're definitely on `splink==4.0.0` and not one of the earlier dev releases?

You can check with 
```
import splink
splink.__version__
```

**Reply by aalexandersson** [reply]
Yes, I am on 4.0.0. This is the output:

`'4.0.0'`

Also, in case it matters, the data type is `object`, so I am not sure how it gets converted to integer.

**Reply by RobinL** [reply]
Thanks for confirming 

With your SSN data:
- is it a dedupe_only, link_only, or link_and_dedupe
-  how many rows do you have (per dataframe if more than one)

That will hopefully allow me to recreate the issue

**Reply by aalexandersson** [reply]
The link type is a `link_only`. Here is the settings configuration:

```
settings = SettingsCreator(                             # <3>
    unique_id_column_name="simulant_id",                # <4>
    link_type="link_only",                              # <5>
    comparisons=[                                       # <7>
        cl.JaroWinklerAtThresholds("first_name", [0.9, 0.7]),
        cl.JaroWinklerAtThresholds("last_name", [0.9, 0.7]),
        cl.LevenshteinAtThresholds("dob", [1,2]),            # 
        cl.LevenshteinAtThresholds("street_number", [1,2]),  #  
        cl.LevenshteinAtThresholds("street_name", [1,2]),
        cl.DamerauLevenshteinAtThresholds("ssn", [1, 2]),
        cl.ExactMatch("zip"),
        cl.ExactMatch("sex"),
    ],
    retain_matching_columns=True,                     # <8>
    retain_intermediate_calculation_columns=True,
    em_convergence=0.0001,                             # <9>
)
```
`df1` has 670251 rows and `df2` has 670214 rows. Below is more sample output.

```
df1.ssn
df2.ssn
```

```
0         449632668
1         349511166
2         023141049
3         553530413
4         048133388
            ...    
670246    746654363
670247    483701187
670248          NaN
670249    308590487
670250    240266261
Name: ssn, Length: 670251, dtype: object
0         449632668
1         349518166
2               NaN
3         553530413
4         718020331
            ...    
670209    746654363
670210    483701187
670211    032571364
670212    308590487
670213    240Z66261
Name: ssn, Length: 670214, dtype: object
```

**Reply by RobinL** [reply]
Hmm - this code runs successfully for me with the following output.  Can you see any way it's obviously different from yours?  Does it run successfully at your end?

```
Probability two random records match is estimated to be  9.87e-07.
This means that amongst all possible pairwise record comparisons, one in 1,012,846.23 are expected to match.  With 449,211,603,714 total possible comparisons, we expect a total of around 443,514.12 matching pairs
num_observed_matches: 376987
num_total_comparisons: 449211603714
recall: 0.85
estimated matches: 381829863156.89996
```

```
import numpy as np
import pandas as pd

from splink import DuckDBAPI, Linker, SettingsCreator, block_on
from splink.blocking_analysis import (
    count_comparisons_from_blocking_rule,
)

ssn_1 = np.random.randint(1, 670251, size=670251)
ssn_2 = np.random.randint(1, 670214, size=670214)

df_1 = pd.DataFrame({"ssn": ssn_1})
df_2 = pd.DataFrame({"ssn": ssn_2})

# create column called unique_id from index
df_1["unique_id"] = df_1.index
df_2["unique_id"] = df_2.index


num_missing_1 = int(len(df_1) * 0.25)
missing_indices_1 = np.random.choice(df_1.index, num_missing_1, replace=False)
df_1.loc[missing_indices_1, "ssn"] = None


num_missing_2 = int(len(df_2) * 0.25)
missing_indices_2 = np.random.choice(df_2.index, num_missing_2, replace=False)
df_2.loc[missing_indices_2, "ssn"] = None


settings = SettingsCreator(link_type="link_only")

db_api = DuckDBAPI()

linker = Linker([df_1, df_2], settings, db_api)

deterministic_rules = [
    block_on("ssn"),
]
linker.training.estimate_probability_two_random_records_match(
    deterministic_rules, recall=0.85
)

```


Another thing to try could be to run the sql that Splink generates against your input data directly.  The SQL to do the count is as follows, which you can see by enabling logging as follows before running your code:

```
linker = Linker([df_1, df_2], settings, db_api)

import logging 
logging.basicConfig(format="%(message)s")
logging.getLogger("splink").setLevel(1)
```


```
CREATE TABLE __splink__df_count_cumulative_blocks_07f6a1693 AS 
WITH

__splink__df_concat as (
            select
            '__splink__input_table_0' as source_dataset,
            "ssn", "unique_id"
            
            from __splink__input_table_0
             UNION ALL 
            select
            '__splink__input_table_1' as source_dataset,
            "ssn", "unique_id"
            
            from __splink__input_table_1
            ), 

__splink__df_concat_left as (
        select * from __splink__df_concat
        where "source_dataset" =
            (select min("source_dataset") from __splink__df_concat)
        ), 

__splink__df_concat_right as (
        select * from __splink__df_concat
        where "source_dataset" =
            (select max("source_dataset") from __splink__df_concat)
        ), 

__splink__blocked_id_pairs as (
            select
            '0' as match_key,
            l."source_dataset" || '-__-' || l."unique_id" as join_key_l,
            r."source_dataset" || '-__-' || r."unique_id" as join_key_r
            from __splink__df_concat_left as l
            inner join __splink__df_concat_right as r
            on
            (l."ssn" = r."ssn")
             where 1=1 
            
            ) 

        select
        count(*) as row_count,
        match_key
        from __splink__blocked_id_pairs
        group by match_key
        order by cast(match_key as int) asc
```

You could run something similar on my test data like:


```
import duckdb

sql = """
with

__splink__blocked_id_pairs as (
select
            '0' as match_key,
            l."unique_id" as join_key_l,
            r."unique_id" as join_key_r
            from df_1 as l
            inner join df_2 as r
            on
            (l."ssn" = r."ssn")
             where 1=1 
)
 select
        count(*) as row_count,
        match_key
        from __splink__blocked_id_pairs
        group by match_key
        order by cast(match_key as int) asc
"""
duckdb.sql(sql)
```

Does that give you a sensible answer int the `row_count` field?

I get:
```
┌───────────┬───────────┐
│ row_count │ match_key │
│   int64   │  varchar  │
├───────────┼───────────┤
│    377818 │ 0         │
└───────────┴───────────┘
```

**Reply by aalexandersson** [reply]
Your test data gives me the same error message on my end:
```
----- stdout -----
num_observed_matches: 376302
num_total_comparisons: -2147483648
recall: 0.85
estimated matches: -1825361100.8
------------------

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [1], in <cell line: 0>()
     34 linker = Linker([df_1, df_2], settings, db_api)
     36 deterministic_rules = [
     37     block_on("ssn"),
     38 ]
---> 39 linker.training.estimate_probability_two_random_records_match(
     40     deterministic_rules, recall=0.85
     41 )

File V:\Testing\Splink_2025\.venv\Lib\site-packages\splink\internals\linker_components\training.py:124, in LinkerTraining.estimate_probability_two_random_records_match(self, deterministic_matching_rules, recall, max_rows_limit)
    121 print(f"estimated matches: {num_total_comparisons * recall}")
    123 if num_observed_matches > num_total_comparisons * recall:
--> 124     raise ValueError(
    125         f"Deterministic matching rules led to more "
    126         f"observed matches than is consistent with supplied recall. "
    127         f"With these rules, recall must be at least "
    128         f"{num_observed_matches/num_total_comparisons:,.2f}."
    129     )
    131 num_expected_matches = num_observed_matches / recall
    132 prob = num_expected_matches / num_total_comparisons

ValueError: Deterministic matching rules led to more observed matches than is consistent with supplied recall. With these rules, recall must be at least -0.00.
```
In case it helps, I attach my [requirements.txt](https://github.com/user-attachments/files/16603275/requirements.txt) file.

Your SQL code on your test data works fine on my end. The is the `row_count` output:
```
┌───────────┬───────────┐
│ row_count │ match_key │
│   int64   │  varchar  │
├───────────┼───────────┤
│    377511 │ 0         │
└───────────┴───────────┘
```

Your SQL code on my test data (with data frames `df1` and `df2` rather than `df_1` and `df_2`) results in multiple syntax errors. Maybe some simple typo, wrong linebreak or similar beginner copy-paste error somewhere in my code?
```
>>> CREATE TABLE __splink__df_count_cumulative_blocks_07f6a1693 AS 
  File "<string>", line 1
    CREATE TABLE __splink__df_count_cumulative_blocks_07f6a1693 AS
           ^^^^^
SyntaxError: invalid syntax
>>> WITH
Traceback (most recent call last):
  File "<string>", line 1, in <module>
NameError: name 'WITH' is not defined
>>> 
>>> __splink__df_concat as (
  File "<string>", line 1
    __splink__df_concat as (
                        ^^
SyntaxError: invalid syntax
```

**Reply by RobinL** [reply]
Thanks, I'll take a look again tomorrow.  At least we have something close to a a reprex now.  Thank for the requirements file that'll be very helpful in testing this. What version of python are you running?  Tomorrow I'll try your requirements on Mac and if I still can't reproduce I'll try on Windows. Hopefully getting close to pinning this down now

**Reply by aalexandersson** [reply]
I use Python 3.11.0. The original data are the `pseudopeople` data from Abraham Flaxman, which I accessed [here](https://github.com/ihmeuw/pseudopeople/issues/205). If needed, I could ask him for permission to share the derived data frames `df1` and `df2` with you privately.

**Reply by RobinL** [reply, marked as answer]
I have tried my test script above (the one that creates `ssn_1 = np.random.randint(1, 670251, size=670251)` etc.) on
- Mac OS Sonoma 14.3, Python 3.11.8, with your `requirements.txt`.
- Windows 11, Python 3.10, with a vanilla duckdb install. (I didn't use the whole of `requirements.txt` because it's pretty heavyweight and there's not much space on my old windows laptop)

I can't replicate the error on either

What happens if you create a fresh venv, install only splink, and try the test script.  Do you still get the same error?  How about on a different version of Python?  

A bit stumped on this one!

If you run the test script in VS code debugger (Debug -> Start Debugging),  then it'd be useful narrow down there the negative number first appears

**Reply by aalexandersson** [reply]
The VS Code debugger (on the existing venv) seems not to add any information, see the screenshot. I used VS Code version 1.92.1 (current) for Windows:

![image](https://github.com/user-attachments/assets/429400a8-03a8-4230-a581-451b47cff298)

A fresh install on another Windows 10 computer of Python `3.11.9` with splink, numpy and pandas in `venv` ran your test script fine. Finally some progress on my end:

![image](https://github.com/user-attachments/assets/c063bfd6-4620-462a-a340-161e9bf13436)

Update: The issue has been resolved on my main work computer using my test data. I updated Python from 3.11.0 to 3.11.9, which did not solve the issue. However, a fresh `venv` solved the issue. I attach the updated [requirements.txt](https://github.com/user-attachments/files/16619712/requirements.txt). 

I can no longer save offline charts to PNG for PDF but I will open a separate [issue 2335 (now resolved)](https://github.com/moj-analytical-services/splink/discussions/2335) for that. Thanks for the help!

**Reply by fscholes** [reply]
Not really to add a solution but I was also experiencing this on an environment that previously upgraded from Splink 3 to Splink 4, and the issue was also fixed with a fresh venv. I am running Python 3.12. It would be interesting to see if this affects other IDEs, not just VS Code on Windows (but I use VS Code myself)

**Reply by RobinL** [reply]
Thanks both for the updates and the solution!

**Reply by aalexandersson** [reply]
The issue affected both my IDEs: RStudio and VS Code.  I'm guessing instead that the dependency issue is related to **`numpy`**. Previously, I used `numpy==1.26.4` because `pseudopeople==1.1.1` requires `numpy<2.0.0`. In the fresh `venv` I have `numpy==2.0.1` but no longer `pseudopeople` installed. I will try to find time to test this.

**Reply by aalexandersson** [reply]
I have now done some further testing. On Robin's test data, the use of current `pandas==2.2.2` on line 12 creates the error `TypeError: Cannot convert numpy.ndarray to numpy.ndarray` when the `numpy` dependency is the old `numpy==1.26.4`:

```
TypeError                                 Traceback (most recent call last)
Cell In[1], line 12
      9 ssn_1 = np.random.randint(1, 670251, size=670251)
     10 ssn_2 = np.random.randint(1, 670214, size=670214)
---> 12 df_1 = pd.DataFrame({"ssn": ssn_1})
     13 df_2 = pd.DataFrame({"ssn": ssn_2})
     15 # create column called unique_id from index

File V:\Testing\Splink_2025\.venv\Lib\site-packages\pandas\core\frame.py:778, in DataFrame.__init__(self, data, index, columns, dtype, copy)
    772     mgr = self._init_mgr(
    773         data, axes={"index": index, "columns": columns}, dtype=dtype, copy=copy
    774     )
    776 elif isinstance(data, dict):
    777     # GH#38939 de facto copy defaults to False only in non-dict cases
--> 778     mgr = dict_to_mgr(data, index, columns, dtype=dtype, copy=copy, typ=manager)
    779 elif isinstance(data, ma.MaskedArray):
    780     from numpy.ma import mrecords

File V:\Testing\Splink_2025\.venv\Lib\site-packages\pandas\core\internals\construction.py:480, in dict_to_mgr(data, index, columns, dtype, typ, copy)
    478 else:
    479     keys = list(data.keys())
--> 480     columns = Index(keys) if keys else default_index(0)
    481     arrays = [com.maybe_iterable_to_list(data[k]) for k in keys]
    483 if copy:

File V:\Testing\Splink_2025\.venv\Lib\site-packages\pandas\core\indexes\base.py:565, in Index.__new__(cls, data, dtype, copy, name, tupleize_cols)
    562         data = com.asarray_tuplesafe(data, dtype=_dtype_obj)
    564 try:
--> 565     arr = sanitize_array(data, None, dtype=dtype, copy=copy)
    566 except ValueError as err:
    567     if "index must be specified when data is not list-like" in str(err):

File V:\Testing\Splink_2025\.venv\Lib\site-packages\pandas\core\construction.py:654, in sanitize_array(data, index, dtype, copy, allow_2d)
    651     subarr = _try_cast(data, dtype, copy)
    653 else:
--> 654     subarr = maybe_convert_platform(data)
    655     if subarr.dtype == object:
    656         subarr = cast(np.ndarray, subarr)

File V:\Testing\Splink_2025\.venv\Lib\site-packages\pandas\core\dtypes\cast.py:139, in maybe_convert_platform(values)
    137 if arr.dtype == _dtype_obj:
    138     arr = cast(np.ndarray, arr)
--> 139     arr = lib.maybe_convert_objects(arr)
    141 return arr

File lib.pyx:2538, in pandas._libs.lib.maybe_convert_objects()

TypeError: Cannot convert numpy.ndarray to numpy.ndarray
```
I am allowed to share data frames `df1` and `df2` with Robin privately if needed, but it seems that this is not needed. 

I conclude that `splink==4.0.0` and `pandas==2.2.2` work with `numpy==2.0.1` whereas `pseudopeople==1.1.1` do not.

**Reply by aflaxman** [reply]
Thanks for using `pseudopeople` in this work!  In our next release, I expect we will be able to install with `numpy` v2.0, but we currently have an internal dependence on `vivarium` which will need more work before it is ready for `numpy` v2.0.  I don't think `pseudopeople` is actually using any of the parts of `vivarium` that will break with the newer `numpy`, though, so if you want an adventure, you can try running `pseudopeople` with `numpy==2.0.1` and see if it works.  If it fails, I would be interested to know why!  (But if you've got to a point where you are ready to move on then I can definitely understand that!)

**Reply by LeCodeMinister** [reply]
> I have now done some further testing. On Robin's test data, the use of current `pandas==2.2.2` on line 12 creates the error `TypeError: Cannot convert numpy.ndarray to numpy.ndarray` when the `numpy` dependency is the old `numpy==1.26.4`:
> 
> ```
> TypeError                                 Traceback (most recent call last)
> Cell In[1], line 12
>       9 ssn_1 = np.random.randint(1, 670251, size=670251)
>      10 ssn_2 = np.random.randint(1, 670214, size=670214)
> ---> 12 df_1 = pd.DataFrame({"ssn": ssn_1})
>      13 df_2 = pd.DataFrame({"ssn": ssn_2})
>      15 # create column called unique_id from index
> 
> File V:\Testing\Splink_2025\.venv\Lib\site-packages\pandas\core\frame.py:778, in DataFrame.__init__(self, data, index, columns, dtype, copy)
>     772     mgr = self._init_mgr(
>     773         data, axes={"index": index, "columns": columns}, dtype=dtype, copy=copy
>     774     )
>     776 elif isinstance(data, dict):
>     777     # GH#38939 de facto copy defaults to False only in non-dict cases
> --> 778     mgr = dict_to_mgr(data, index, columns, dtype=dtype, copy=copy, typ=manager)
>     779 elif isinstance(data, ma.MaskedArray):
>     780     from numpy.ma import mrecords
> 
> File V:\Testing\Splink_2025\.venv\Lib\site-packages\pandas\core\internals\construction.py:480, in dict_to_mgr(data, index, columns, dtype, typ, copy)
>     478 else:
>     479     keys = list(data.keys())
> --> 480     columns = Index(keys) if keys else default_index(0)
>     481     arrays = [com.maybe_iterable_to_list(data[k]) for k in keys]
>     483 if copy:
> 
> File V:\Testing\Splink_2025\.venv\Lib\site-packages\pandas\core\indexes\base.py:565, in Index.__new__(cls, data, dtype, copy, name, tupleize_cols)
>     562         data = com.asarray_tuplesafe(data, dtype=_dtype_obj)
>     564 try:
> --> 565     arr = sanitize_array(data, None, dtype=dtype, copy=copy)
>     566 except ValueError as err:
>     567     if "index must be specified when data is not list-like" in str(err):
> 
> File V:\Testing\Splink_2025\.venv\Lib\site-packages\pandas\core\construction.py:654, in sanitize_array(data, index, dtype, copy, allow_2d)
>     651     subarr = _try_cast(data, dtype, copy)
>     653 else:
> --> 654     subarr = maybe_convert_platform(data)
>     655     if subarr.dtype == object:
>     656         subarr = cast(np.ndarray, subarr)
> 
> File V:\Testing\Splink_2025\.venv\Lib\site-packages\pandas\core\dtypes\cast.py:139, in maybe_convert_platform(values)
>     137 if arr.dtype == _dtype_obj:
>     138     arr = cast(np.ndarray, arr)
> --> 139     arr = lib.maybe_convert_objects(arr)
>     141 return arr
> 
> File lib.pyx:2538, in pandas._libs.lib.maybe_convert_objects()
> 
> TypeError: Cannot convert numpy.ndarray to numpy.ndarray
> ```
> 
> I am allowed to share data frames `df1` and `df2` with Robin privately if needed, but it seems that this is not needed.
> 
> I conclude that `splink==4.0.0` and `pandas==2.2.2` work with `numpy==2.0.1` whereas `pseudopeople==1.1.1` do not.

This has saved me today.

I had the ValueError on my work machine because I was using `pandas==2.2.1` and `numpy==2.0.2`. `pandas==2.2.1` requires `numpy<2,>1.22.4`, which has caused the error. Updating Pandas to `2.2.2` has solved the ValueError. Thank you!

**Reply by aalexandersson** [reply]
Yes, `pandas 2.2.2` is the first version of pandas that is generally compatible with `numpy 2.x`. It is documented at [pandas.pydata.org/docs/whatsnew/v2.2.2.html#pandas-2-2-2-is-now-compatible-with-numpy-2-0](https://pandas.pydata.org/docs/whatsnew/v2.2.2.html#pandas-2-2-2-is-now-compatible-with-numpy-2-0).
